### PR TITLE
NSPR-1387 : Xrt hbm support

### PR DESCRIPTION
**Pull request by saifuddin-xilinx**
Hi,
Creating this PR for review this HBM changes.
Please review the changes follows:

Userspace changes for XRT layer
User_PF Driver changes for HBM support
New changes support the following functionality:

User can specify a range of HBM banks in the v++ connectivity section, Example:
-sp CU_0.in1:HBM[0:3] . The user needs to specify bank-id information in his host application
as well using Extension Pointer.
If the user doesn't specify anything through the Extension pointer, Then XRT should automatically
detect the HBM banks using the connectivity section present in rtd file.

Further details : https://confluence.xilinx.com/pages/viewpage.action?pageId=205359608

### CODE REVIEW

File: src/runtime_src/core/pcie/linux/shim.cpp [hunk 1]
```diff
@@ -2235,8 +2275,18 @@ int xclLoadXclBin(xclDeviceHandle handle, const xclBin *buffer)
...
 
     if (!ret) {
+      char *pMemInfo = NULL;
       auto core_device = xrt_core::get_userpf_device(drv);
-      core_device->register_axlf(buffer);
+      
+      /* Get memory topology info from driver through sysfs */
+      drv->xclGetMemInfo(&pMemInfo);
+      /* Populate the retrive info into device class */
+      core_device->register_axlf(buffer, pMemInfo);
```
Comment (stsoe) [inline src/runtime_src/core/pcie/linux/shim.cpp:2284]: What is pMemInfo?  All the sections are part of the xclbin or should be accessed via query requests, if group_topology/connectivity are not in the xclbin or not in xclbin, then register_axlf should simply copy mem_topology and connectivity.  I don't think register_axlf signature should change.

File: src/runtime_src/core/pcie/linux/shim.h [hunk 2]
```diff
@@ -83,6 +83,7 @@ class shim
     int xclExportBO(unsigned int boHandle);
     unsigned int xclImportBO(int fd, unsigned flags);
     int xclGetBOProperties(unsigned int boHandle, xclBOProperties *properties);
+    int xclGetMemInfo(char **MemInfo);
```
Comment (stsoe) [inline src/runtime_src/core/pcie/linux/shim.h:86]: What is this function?

File: src/runtime_src/core/common/device.cpp [hunk 3]
```diff
@@ -62,18 +62,77 @@ get_xclbin_uuid() const
...
+  }
+}
+
+void
+device::
+register_axlf(const axlf* top, const char *info_buff)
 {
   m_axlf_sections.clear();
   m_xclbin_uuid = uuid(top->m_header.uuid);
   axlf_section_kind kinds[] = {EMBEDDED_METADATA, AIE_METADATA, IP_LAYOUT, CONNECTIVITY, MEM_TOPOLOGY, DEBUG_IP_LAYOUT, SYSTEM_METADATA};
```
Comment (stsoe) [inline src/runtime_src/core/common/device.cpp:112]: Please expand to include GROUP_TOPOLOGY and GROUP_CONNECTIVITY.  The conditional (if kind) code should chekc for if kind is GROUP_TOPOLOGY or GROUP_CONNECTIVITY and these are not in xclbin, in which case they should just be copies of MEM_TOPOLOGY and CONNECTIVITY

File: src/runtime_src/core/common/device.cpp [hunk 4]
```diff
@@ -62,18 +62,77 @@ get_xclbin_uuid() const
 
 void
 device::
-register_axlf(const axlf* top)
+get_section_info(const char *buff, axlf_section_kind kind, 
+        const char *&sect_info, size_t *sect_size)
```
Comment (stsoe) [inline src/runtime_src/core/common/device.cpp:66]: I am not understand the need for this function?

File: src/runtime_src/core/common/device.h [hunk 5]
```diff
@@ -170,14 +170,23 @@ class device : public ishim
...
 
+  /**
+   * get_section_info() - Get the section info from the input buffer  
+   * received from Driver
+   *
+   */
+  void 
+  get_section_info(const char *buff, axlf_section_kind kind,
+                const char *&sect_info, size_t *sect_size);
+
```
Comment (stsoe) [inline src/runtime_src/core/common/device.h:181]: Not sure this function is needed?

File: src/runtime_src/core/common/device.h [hunk 6]
```diff
@@ -170,14 +170,23 @@ class device : public ishim
...
   /**
    * register_axlf() - Callback from shim after AXLF has been loaded.
    *
    * This function extracts meta data sections as needed.
    */
   XRT_CORE_COMMON_EXPORT
   void
-  register_axlf(const axlf*);
+  register_axlf(const axlf*, const char *info_buff = nullptr);
```
Comment (stsoe) [inline src/runtime_src/core/common/device.h:189]: I don't think the signature should include info_buff.?

File: src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_drm.c [hunk 7]
```diff
@@ -683,14 +731,17 @@ int xocl_cleanup_mem(struct xocl_drm *drm_p)
...
 	struct mem_topology *topo = NULL;
+	struct mem_topology *group_topo = NULL;
 	struct mem_data *mem_data;
 	uint32_t shared;
-	struct xocl_mm_wrapper *wrapper = NULL;
-	uint64_t reserved1 = 0;
+    struct xocl_mm_wrapper *wrapper = NULL;
+    uint64_t mm_start_addr = 0;
+    uint64_t mm_end_addr = 0;
+    uint64_t reserved1 = 0;
```
Comment (maxzhen) [inline src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_drm.c:744]: bad indentation

File: src/runtime_src/core/pcie/linux/shim.cpp (hunk 1 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/pcie/linux/shim.cpp]: Removed this function.

File: src/runtime_src/core/pcie/linux/shim.h (hunk 2 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/pcie/linux/shim.h]: Now this function is not required. Getting the group info from the xclbin directly.

File: src/runtime_src/core/common/device.cpp (hunk 3 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/common/device.cpp]: Updated the function.

File: src/runtime_src/core/common/device.h (hunk 5 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/common/device.h]: Removed.

File: src/runtime_src/core/common/device.h (hunk 6 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/common/device.h]: Reverted to the original signature.

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c [hunk 8]
```diff
@@ -2359,6 +2375,68 @@ static void icap_get_max_host_mem_aperture(struct icap *icap)
 	return;
 }
 
+static int init_memory_group(struct platform_device *pdev, struct axlf *xclbin)
```
[the remark this answers is not in the document]
Comment (saifuddin-xilinx) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/icap.c]: Updated as per your comments.

File: src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_drm.c [hunk 9]
```diff
@@ -520,14 +520,47 @@ void xocl_mm_update_usage_stat(struct xocl_drm *drm_p, u32 ddr,
 	drm_p->mm_usage_stat[ddr]->bo_count += count;
 }
 
+int xocl_mm_insert_node_range(struct xocl_drm *drm_p, u32 mem_id,
+			      struct drm_mm_node *node, u64 size)
+{
+	uint64_t start_addr = 0;
+	uint64_t end_addr = 0;
+	struct mem_topology *grp_topology = NULL;
+    int err = 0;
```
[the remark this answers is not in the document]
Comment (saifuddin-xilinx) [reply src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_drm.c]: Fixed.

File: src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_drm.c (hunk 7 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_drm.c]: Fixed.

File: src/runtime_src/core/common/device.cpp (hunk 4 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/common/device.cpp]: Removed this.. Not required now.

File: src/runtime_src/core/common/device.cpp [hunk 10]
```diff
@@ -66,11 +66,22 @@ register_axlf(const axlf* top)
...
+  axlf_section_kind kinds[] = {EMBEDDED_METADATA, AIE_METADATA, IP_LAYOUT, CONNECTIVITY, 
+			ASK_GROUP_CONNECTIVITY, ASK_GROUP_TOPOLOGY, MEM_TOPOLOGY, DEBUG_IP_LAYOUT, SYSTEM_METADATA};
   for (auto kind : kinds) {
     auto hdr = ::xclbin::get_axlf_section(top, kind);
-    if (!hdr)
-      continue;
+    if (!hdr) {
+      /* If group topology or group connectivity doesn't exists then use the
+       * mem topology or connectivity respectively section for the same.
+       */
```
Comment (stsoe) [inline src/runtime_src/core/common/device.cpp:76]: Good comment.  Minor, please use C++ comment style  //

**Comment by stsoe** [comment]
I am marking this as not merge .  I want to run some testing just as sanity check first.

**Comment by stsoe** [comment]
HW tests passed

**Comment by maxzhen** [comment]
@chienwei-lan , please also try it with slave bridge test cases.

File: src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_drm.c [hunk 11]
```diff
@@ -520,14 +520,46 @@ void xocl_mm_update_usage_stat(struct xocl_drm *drm_p, u32 ddr,
...
 }
 
+int xocl_mm_insert_node_range(struct xocl_drm *drm_p, u32 mem_id,
+			      struct drm_mm_node *node, u64 size)
+{
+	uint64_t start_addr = 0;
+	uint64_t end_addr = 0;
+	struct mem_topology *grp_topology = NULL;
+	int err = 0;
+	BUG_ON(!mutex_is_locked(&drm_p->mm_lock));
```
Comment (rozumx) [inline src/runtime_src/core/pcie/driver/linux/xocl/userpf/xocl_drm.c:530]: Hmm...what does this macro "BUG_ON" do?

**Comment by chvamshi-xilinx** [comment]
> 
> 
> There seems to be some missing functionality. I see changes for xbutil, but not xbutil2 (e.g., core/tools/xbutil2). **_Is this an oversight?_**
> 
> _Question_
> If an older xclbin image is loaded with NO GROUP_TOPOLOGY or GROUP_CONNECTIVITY, what will be the expected behavior (e.g., same as 2020.1 or this use case is no longer expected)?

Behavior would be same as 2020.1. It is supported and validated this functionality.

Changes to xbutil2 are in the plan.  We will add group_topology information in xbutil2 once it has an interface to show mem_topology . Currently it is not available

**Comment by stsoe** [comment]
@chvamshi-xilinx @rozumx Can we merge this PR?   I have pending changes to switch user space to use the group sections.

**Comment by chvamshi-xilinx** [comment]
> 
> 
> @chvamshi-xilinx @rozumx Can we merge this PR? I have pending changes to switch user space to use the group sections.

I am fine to merge this change. xbutil2 changes can be done once the interface is available.

**Pull request merged**